### PR TITLE
Use MessagePackReader \ MessagePackWriter to implement IHubProtocol for server-side Blazor

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/7311

### CODE REVIEW

File: src/Components/Server/src/Circuits/RemoteRenderer.cs [hunk 1]
```diff
@@ -112,7 +113,13 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
             // snapshot its contents now.
             // TODO: Consider using some kind of array pool instead of allocating a new
             //       buffer on every render.
-            var batchBytes = MessagePackSerializer.Serialize(batch, RenderBatchFormatterResolver.Instance);
+            var memoryStream = new MemoryStream();
+            using (var renderBatchWriter = new RenderBatchWriter(memoryStream, false))
```
Comment (BrennanConroy) [inline src/Components/Server/src/Circuits/RemoteRenderer.cs]: Hope there are some perf improvement possibilities here :D

File: src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,434 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
```
Comment (BrennanConroy) [inline src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTestBase.cs]: This makes me a bit sad.

Add a comment explaining that this is shared with Razor Components and when adding tests put it in MessagePackHubProtocolTests.cs unless they are simple tests.

**Review by BrennanConroy — COMMENTED** [review]
Talk to @Eilon about the submodule/code usage

File: .gitmodules [hunk 3]
```diff
@@ -1,3 +1,7 @@
 [submodule "googletest"]
 	path = src/submodules/googletest
 	url = https://github.com/google/googletest
+
+[submodule "MessagePack"]
+	path = src/submodules/MessagePack
+	url = https://github.com/pranavkm/MessagePack-CSharp
```
Comment (Eilon) [inline .gitmodules]: We have our own fork now: https://github.com/aspnet/MessagePack-CSharp

File: src/Components/Server/src/Circuits/RemoteRenderer.cs [hunk 4]
```diff
@@ -112,7 +113,13 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
...
+            {
+                renderBatchWriter.Write(in batch);
+            }
+
+            var batchBytes = memoryStream.ToArray();
```
Comment (davidfowl) [inline src/Components/Server/src/Circuits/RemoteRenderer.cs]: You can avoid the second array allocation if you supported Memory\<byte\> or ArraySegment\<byte\>

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+
+            var target = ReadString(ref reader, "target");
+
+            object[] arguments;
+            try
+            {
+                var parameterTypes = binder.GetParameterTypes(target);
+                arguments = BindArguments(ref reader, parameterTypes);
+            }
+            catch (Exception ex) when (!(ex is FormatException))
```
Comment (BrennanConroy) [inline src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: catch all

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+            var invocationId = ReadString(ref reader, "invocationId");
+            var target = ReadString(ref reader, "target"); ;
+
+            object[] arguments;
+            try
+            {
+                var parameterTypes = binder.GetParameterTypes(target);
+                arguments = BindArguments(ref reader, parameterTypes);
+            }
+            catch (Exception ex) when (!(ex is FormatException))
```
Comment (BrennanConroy) [inline src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: catch all

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+                {
+                    return reader.ReadInt64();
+                }
+                else if (type == typeof(float))
+                {
+                    return reader.ReadSingle();
+                }
+                else if (type == typeof(byte[]))
+                {
+                    var bytes = reader.ReadBytes();
```
Comment (BrennanConroy) [inline src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: We should be able to `ToArray` this

File: src/Components/Server/src/Circuits/RemoteRenderer.cs [hunk 8]
```diff
@@ -112,7 +113,13 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
             // snapshot its contents now.
             // TODO: Consider using some kind of array pool instead of allocating a new
             //       buffer on every render.
-            var batchBytes = MessagePackSerializer.Serialize(batch, RenderBatchFormatterResolver.Instance);
+            var memoryStream = new MemoryStream();
```
Comment (BrennanConroy) [inline src/Components/Server/src/Circuits/RemoteRenderer.cs]: nit: using var

File: src/Components/Server/test/Microsoft.AspNetCore.Components.Server.Tests.csproj [hunk 9]
```diff
@@ -11,8 +11,17 @@
     <Reference Include="Microsoft.Extensions.Logging.Testing" />
   </ItemGroup>
 
+  <PropertyGroup>
+    <SignalRTestBase>$(RepositoryRoot)src\SignalR\common\SignalR.Common\test\Internal\Protocol\</SignalRTestBase>
+  </PropertyGroup>
+
   <ItemGroup>
     <Compile Include="..\..\Components\test\Rendering\HtmlRendererTestBase.cs" />
+    <Compile Include="$(SignalRTestBase)CustomObject.cs" LinkBase="BlazorPack"/>
```
Comment (BrennanConroy) [inline src/Components/Server/test/Microsoft.AspNetCore.Components.Server.Tests.csproj]: Review the shared files, for example CustomObject.cs shouldn't be needed

File: src/Components/Server/src/BlazorPack/NativeDateTimeFormatter.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace MessagePack.Formatters
+{
+    internal class NativeDateTimeFormatter
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Components/Server/src/BlazorPack/NativeDateTimeFormatter.cs]: We'll clean this up eventually. Right now it's required for the code to compile.

File: src/Components/Server/src/Circuits/RemoteRenderer.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/Server/src/Circuits/RemoteRenderer.cs]: There's a TODO there which we'll get to at some point. Right now the focus is to get something in.

File: src/Components/Server/src/Circuits/RemoteRenderer.cs (hunk 4 above)
Comment (pranavkm) [reply src/Components/Server/src/Circuits/RemoteRenderer.cs]: Would this be in SignalR?

**Comment by pranavkm** [comment]
🆙 📅

File: src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTests.cs [hunk 11]
```diff
@@ -52,34 +86,41 @@ public static IEnumerable<object[]> TestDataNames
...
-            new ProtocolTestData(
-                name: "InvocationWithNoHeadersNoIdIntAndStringArgs",
-                message: new InvocationMessage("method", new object[] { 42, "string" }),
-                binary: "lgGAwKZtZXRob2SSKqZzdHJpbmeQ"),
             new ProtocolTestData(
                 name: "InvocationWithNoHeadersNoIdIntAndEnumArgs",
                 message: new InvocationMessage("method", new object[] { 42, TestEnum.One }),
                 binary: "lgGAwKZtZXRob2SSKqNPbmWQ"),
-            new ProtocolTestData(
+             new ProtocolTestData(
```
Comment (BrennanConroy) [inline src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTests.cs]: nit: remove space

File: src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTests.cs [hunk 12]
```diff
@@ -52,34 +86,41 @@ public static IEnumerable<object[]> TestDataNames
...
+                binary: "lQODo0Zvb6NCYXKyS2V5V2l0aApOZXcNCkxpbmVzq1N0aWxsIFdvcmtzsVZhbHVlV2l0aE5ld0xpbmVzsEFsc28KV29ya3MNCkZpbmWjeHl6A5KGqlN0cmluZ1Byb3CoU2lnbmFsUiGqRG91YmxlUHJvcMtAGSH7VELPEqdJbnRQcm9wKqxEYXRlVGltZVByb3DW/1jsHICoTnVsbFByb3DAq0J5dGVBcnJQcm9wxAMBAgOGqlN0cmluZ1Byb3CoU2lnbmFsUiGqRG91YmxlUHJvcMtAGSH7VELPEqdJbnRQcm9wKqxEYXRlVGltZVByb3DW/1jsHICoTnVsbFByb3DAq0J5dGVBcnJQcm9wxAMBAgM="),
             new ProtocolTestData(
-                name: "InvocationWithNoHeadersNoIdAndNoArgs",
-                message: new InvocationMessage("method", Array.Empty<object>()),
-                binary: "lgGAwKZtZXRob2SQkA=="),
+                name: "CompletionWithNoHeadersAndEnumResult",
+                message: CompletionMessage.WithResult("xyz", payload: TestEnum.One),
+                binary: "lQOAo3h5egOjT25l"),
+
+             // Invocation messages
```
Comment (BrennanConroy) [inline src/SignalR/common/SignalR.Common/test/Internal/Protocol/MessagePackHubProtocolTests.cs]: nit: remove space

**Review by BrennanConroy — APPROVED** [review]

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+using Microsoft.AspNetCore.Internal;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.Components.Server.BlazorPack
+{
+    /// <summary>
+    /// Implements the SignalR Hub Protocol using MessagePack with limited type support. 
+    /// </summary>
+    public class BlazorPackHubProtocol : IHubProtocol
```
Comment (halter73) [inline src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: Do we expect there will be consumers other than Blazor for this?

File: src/Components/Browser.JS/src/Boot.Server.ts [hunk 14]
```diff
@@ -42,9 +42,11 @@ async function boot() {
 }
 
 async function initializeConnection(circuitHandlers: CircuitHandler[]): Promise<signalR.HubConnection> {
+  const hubProtocol = new MessagePackHubProtocol();
+  (hubProtocol as any).name = 'blazorpack';
```
Comment (rynowak) [inline src/Components/Browser.JS/src/Boot.Server.ts]: 😻

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs (hunk 13 above)
Comment (rynowak) [reply src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: Is there a super duper good reason this is public?

Comment (pranavkm) [reply src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: Nope. Just accidental.

**Comment by pranavkm** [comment]
@aspnet-hello triage https://github.com/aspnet/AspNetCore/pull/8687/checks?check_run_id=90265125

**Comment by BrennanConroy** [comment]
Wrong link

**Comment by BrennanConroy** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=139471&view=logs&jobId=2fe95fc5-ae3d-58fd-c2b8-46bb81541875

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact ryanbrandenburg.

I've triaged the above build.

**Comment by Eilon** [comment]
Hold on, I didn't sign off on this. We need 3rd Party Notices for this.

**Comment by pranavkm** [comment]
@Eilon oops. Should I back out the change?

**Comment by Eilon** [comment]
Naah it's fine, I'll let you know what to add.

File: src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs (hunk 13 above)
Comment (halter73) [reply src/Components/Server/src/BlazorPack/BlazorPackHubProtocol.cs]: That's why I asked that question. Guess I should have been more explicit.

**Pull request merged**